src/main/java/com/faforever/client/fx/BrowserCallback.java

```diff
@@ -0,0 +1,223 @@
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class BrowserCallback {
+  private final PlatformService platformService;
+  private final UrlPreviewResolver urlPreviewResolver;
+  private final ReplayService replayService;
+  private final EventBus eventBus;
+  private final ExternalReplayInfoGenerator externalReplayInfoGenerator;
+  private final Pattern replayUrlPattern;
+  private final ClanService clanService;
+  private final UiService uiService;
+  private final PlayerService playerService;
+  private final I18n i18n;
+  @VisibleForTesting
+  Popup clanInfoPopup;
+  private Tooltip linkPreviewTooltip;
+  private Popup playerInfoPopup;
+  private double lastMouseX;
+  private double lastMouseY;
+
+  BrowserCallback(PlatformService platformService, ClientProperties clientProperties,
+                  UrlPreviewResolver urlPreviewResolver, ReplayService replayService, EventBus eventBus,
+                  ExternalReplayInfoGenerator externalReplayInfoGenerator, ClanService clanService, UiService uiService, PlayerService playerService, I18n i18n) {
+    this.platformService = platformService;
+    this.urlPreviewResolver = urlPreviewResolver;
+    this.replayService = replayService;
+    this.eventBus = eventBus;
+    this.externalReplayInfoGenerator = externalReplayInfoGenerator;
+    this.clanService = clanService;
+    this.uiService = uiService;
+    this.playerService = playerService;
+    this.i18n = i18n;
+
+    String urlFormat = clientProperties.getVault().getReplayDownloadUrlFormat();
+    String[] splittedFormat = urlFormat.split("%s");
+    replayUrlPattern = Pattern.compile(Pattern.quote(splittedFormat[0]) + "(\\d+)" + Pattern.compile(splittedFormat.length == 2 ? splittedFormat[1] : ""));
+  }
+
+  /**
+   * Called from JavaScript when user clicked a URL.
+   */
+  public void openUrl(String url) {
+    Matcher replayUrlMatcher = replayUrlPattern.matcher(url);
+    if (!replayUrlMatcher.matches()) {
+      platformService.showDocument(url);
+      return;
+    }
+
+    String replayId = replayUrlMatcher.group(1);
+
+    replayService.findById(Integer.parseInt(replayId))
+        .thenAccept(replay -> Platform.runLater(() -> externalReplayInfoGenerator.showExternalReplayInfo(replay, replayId)));
+  }
+
+  /**
+   * Called from JavaScript when user clicks on user name in chat
+   */
+  public void openPrivateMessageTab(String username) {
+    eventBus.post(new InitiatePrivateChatEvent(username));
+  }
+
+  /**
+   * Called from JavaScript when user no longer hovers over an URL.
+   */
+  public void hideUrlPreview() {
+    if (linkPreviewTooltip != null) {
+      linkPreviewTooltip.hide();
+      linkPreviewTooltip = null;
+    }
+  }
+
+  /**
+   * Called from JavaScript when user hovers over an URL.
+   */
+  public void previewUrl(String urlString) {
+    urlPreviewResolver.resolvePreview(urlString)
+        .thenAccept(optionalPreview -> {
+          if (!optionalPreview.isPresent()) {
```

Comment:
You don't wanna call optionalPreview.ifPresent()?  Makes this thing more readable. In the hideUrlPrview method 10 lines above you did the opposite. Due to number of statements?